torchacc/dist/fsdp.py

```diff
@@ -202,3 +204,151 @@ def clip_grad_norm_(self, max_grad_norm):
+        }
+
+        if rank0_only and self.model.rank == 0:
+            consolidate_optim_state_dict[
+                'param_groups'] = optim_state_param_groups
+        if not rank0_only:
+            consolidate_optim_state_dict[
+                'param_groups'] = optim_state_param_groups
+
+        for layer_state, (layer_name, layer_params) in zip(
+                sharded_optim_state.values(),
+                shard_meta_data['flatten_info'].items()):
+            param_names, param_shapes, param_numels = layer_params
+            # get full_param_names of each layer in optim_state_dict
+            full_names = optim_utils._get_layer_full_names(
+                layer_name, param_names)
+            unflat_state_dict = {fn: {} for fn in full_names}
+
+            if rank0_only and self.model.rank == 0:
+                consolidate_optim_state_dict['param_groups'][0][
+                    'params'].append(full_names)
+            if not rank0_only:
+                consolidate_optim_state_dict[
+                    'param_groups'] = optim_state_param_groups
+
+            for state_name, state_params in layer_state.items():
+                tensor_buffer = optim_utils._all_gather_state(
+                    state_params, self.model)
+                xm.mark_step()
+                # only rank0 save consolidate optim_state_dict
+                if rank0_only and self.model.rank == 0:
+                    _, full_params = optim_utils._unflatten_optim_params(
+                        tensor_buffer, layer_name, param_names, param_shapes,
+                        param_numels)
+
+                    for fn, fp in zip(full_names, full_params):
+                        if cpu_offload:
+                            xm.mark_step()
+                            unflat_state_dict[fn][state_name] = fp.cpu()
+                        else:
+                            unflat_state_dict[fn][state_name] = fp
+
+                # all ranks save consolidate optim_state_dict
+                if not rank0_only:
+                    _, full_params = optim_utils._unflatten_optim_params(
+                        tensor_buffer, layer_name, param_names, param_shapes,
+                        param_numels)
+
+                    for fn, fp in zip(full_names, full_params):
+                        if cpu_offload:
+                            xm.mark_step()
+                            unflat_state_dict[fn][state_name] = fp.cpu()
+                        else:
+                            unflat_state_dict[fn][state_name] = fp
+        consolidate_optim_state_dict['state'] = unflat_state_dict
+
+        return consolidate_optim_state_dict
+
+    def load_optim_state_dict(self,
+                              optim_state_dict: Dict[str, Any],
+                              optim: torch.optim.Optimizer,
+                              full_state_dict: str = 'FULL_STATE_DICT',
+                              rank0_only: bool = True) -> Dict[str, Any]:
```

Comment:
same modification as above and add api docs

Reply:
done